src/utils/registry.ts

```diff
@@ -385,22 +412,34 @@ export class Registry {
       // Remove stale browsers.
       await this._validateInstallationCache(linksDir);
 
-      // Install missing browsers for this package.
-      for (const browserName of browserNames) {
-        const revision = this._revision(browserName);
-        const browserDirectory = this.browserDirectory(browserName);
-        const title = `${browserName} v${revision}`;
-        const downloadFileName = `playwright-download-${browserName}-${hostPlatform}-${revision}.zip`;
-        await downloadBrowserWithProgressBar(title, browserDirectory, this.executablePath(browserName)!, this._downloadURL(browserName), downloadFileName).catch(e => {
-          throw new Error(`Failed to download ${title}, caused by\n${e.stack}`);
-        });
-        await fs.promises.writeFile(markerFilePath(browserDirectory), '');
+      // Install browsers for this package.
+      for (const executable of executables) {
+        if (executable._download)
+          await executable._download();
+        else
+          throw new Error(`ERROR: Playwright does not support installing ${executable.name}`);
       }
     } finally {
       await releaseLock();
     }
   }
 
+  private async _downloadExecutable(descriptor: BrowsersJSONDescriptor, executablePath: string | undefined, downloadURLTemplate: string | undefined, downloadHostEnv: string) {
+    if (!downloadURLTemplate || !executablePath)
+      throw new Error(`ERROR: Playwright does not support ${descriptor.name} on ${hostPlatform}`);
+    const downloadHost =
+        (downloadHostEnv && getFromENV(downloadHostEnv)) ||
+        getFromENV('PLAYWRIGHT_DOWNLOAD_HOST') ||
+        'https://playwright.azureedge.net';
+    const downloadURL = util.format(downloadURLTemplate, downloadHost, descriptor.revision);
+    const title = `${descriptor.name} v${descriptor.revision}`;
+    const downloadFileName = `playwright-download-${descriptor.name}-${hostPlatform}-${descriptor.revision}.zip`;
+    await downloadBrowserWithProgressBar(title, descriptor.dir, executablePath, downloadURL, downloadFileName).catch(e => {
+      throw new Error(`Failed to download ${title}, caused by\n${e.stack}`);
+    });
+    await fs.promises.writeFile(markerFilePath(descriptor.dir), '');
+  }
+
   private async _validateInstallationCache(linksDir: string) {
```

Comment:
nits:
- this doesn't rely on registry instance and can be made static. It'll make it more clear
- `linksDir` can be made static too the same way `registryDirectory` is static.

Reply:
I'll see about that.